mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java

```diff
@@ -50,123 +54,311 @@
+                         Op op,
+                         boolean upsert,
+                         DBCollection collection,
+                         DBObject[] dbObjects,
+                         DocCtx[] inputDocs) {
+        // Operate in batches
+        List<DocInfo> batch=new ArrayList<>(batchSize);
+        for(int i=0;i<dbObjects.length;i++) {
+            DocInfo item=new DocInfo(dbObjects[i],inputDocs[i]);
+            batch.add(item);
+            if(batch.size()>=batchSize) {
+                saveDocs(ctx,op,upsert,collection,batch);
+                batch.clear();
+            }
+        }
+        if(!batch.isEmpty()) {
+            saveDocs(ctx,op,upsert,collection,batch);
+        }
     }
 
-    @Override
-    public void saveDoc(CRUDOperationContext ctx,
-                        Op op,
-                        boolean upsert,
-                        DBCollection collection,
-                        EntityMetadata md,
-                        DBObject dbObject,
-                        DocCtx inputDoc) {
-
-        WriteResult result = null;
-
-        DBObject oldDBObject=null;
-
-        Object id=dbObject.get(MongoCRUDController.ID_STR);
-        if(id==null) {
-            LOGGER.debug("Null _id, looking up the doc using identity fields");
-            Field[] identityFields=md.getEntitySchema().getIdentityFields();
-            Object[] identityFieldValues=fill(dbObject,identityFields);
-            if(!isNull(identityFieldValues)) {
-                DBObject lookupq=getLookupQ(identityFields,identityFieldValues);
-                LOGGER.debug("Lookup query: {}",lookupq);
-                oldDBObject=new FindOneCommand(collection,lookupq,null,maxQueryTimeMS).executeAndUnwrap();
-                LOGGER.debug("Retrieved:{}",oldDBObject);
-                if(oldDBObject!=null)
-                    id=oldDBObject.get(MongoCRUDController.ID_STR);
-                LOGGER.debug("Retrieved id:{}",id);
+    private void saveDocs(CRUDOperationContext ctx,
+                          Op op,
+                          boolean upsert,
+                          DBCollection collection,
+                          List<DocInfo> batch) {       
+        // If this is a save operation, we have to load the existing DB objects
+        if(op==DocSaver.Op.save) {
+            LOGGER.debug("Retrieving existing {} documents for save operation",batch.size());
+            List<BasicDBObject> idQueries=new ArrayList<>(batch.size());
+            for(DocInfo doc:batch) {
+                doc.id=getFieldValues(doc.newDoc,idPaths);
+                if(!isNull(doc.id)) {
+                    idQueries.add(doc.getIdQuery());
+                }
+            }
+            if(!idQueries.isEmpty()) {
+                BasicDBObject retrievalq=new BasicDBObject("$or",idQueries);
+                LOGGER.debug("Existing document retrieval query={}",retrievalq);
+                DBCursor cursor=collection.find(retrievalq,null);
```

Comment:
Closed.

On Fri, May 6, 2016 at 11:30 AM, Naveen Malik <notifications@github.com>
wrote:

> In mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java
> <https://github.com/lightblue-platform/lightblue-mongo/pull/259#discussion_r62361923>
> :
>
> > +                          DBCollection collection,
> > +                          List<DocInfo> batch) {
> > +        // If this is a save operation, we have to load the existing DB objects
> > +        if(op==DocSaver.Op.save) {
> > +            LOGGER.debug("Retrieving existing {} documents for save operation",batch.size());
> > +            List<BasicDBObject> idQueries=new ArrayList<>(batch.size());
> > +            for(DocInfo doc:batch) {
> > +                doc.id=getFieldValues(doc.newDoc,idPaths);
> > +                if(!isNull(doc.id)) {
> > +                    idQueries.add(doc.getIdQuery());
> > +                }
> > +            }
> > +            if(!idQueries.isEmpty()) {
> > +                BasicDBObject retrievalq=new BasicDBObject("$or",idQueries);
> > +                LOGGER.debug("Existing document retrieval query={}",retrievalq);
> > +                DBCursor cursor=collection.find(retrievalq,null);
>
> Should close the cursor. Recommend try-with-resource.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-mongo/pull/259/files/a9fc8b0e5beacdb201ecc3189e3ce88ccf38ad1f#r62361923>
>